lib/reply.js

```diff
@@ -429,24 +431,21 @@ Reply.prototype.getResponseTime = function () {
 // - https://promisesaplus.com/ for the definition of thenable
 // - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then for the signature
 Reply.prototype.then = function (fulfilled, rejected) {
-  if (this.sent) {
-    fulfilled()
-    return
-  }
-
-  eos(this.raw, (err) => {
+  return (this.sent ? Promise.resolve() : promisedEos(this.raw))
+  .catch((err) => {
     // We must not treat ERR_STREAM_PREMATURE_CLOSE as
     // an error because it is created by eos, not by the stream.
     if (err && err.code !== 'ERR_STREAM_PREMATURE_CLOSE') {
-      if (rejected) {
-        rejected(err)
-      } else {
-        this.log && this.log.warn('unhandled rejection on reply.then')
-      }
-    } else {
-      fulfilled()
+      if (rejected) throw err
```

Comment:
Why do you throw? This somehow makes me think that we will get unhandled promise rejections crashing node.

Reply:
> This somehow makes me think that we will get unhandled promise rejections crashing node.

No, this will not happen, it's only being thrown if `rejected` is not undefined, and so, it will be captured by `.then()` set later at line 444.